src/main/java/org/truffleruby/parser/parser/ParserSupport.java

```diff
@@ -1240,20 +1240,55 @@ public ArgsTailHolder new_args_tail(SourceIndexLength position, ListParseNode ke
 
     public ParseNode remove_duplicate_keys(HashParseNode hash) {
         List<ParseNode> encounteredKeys = new ArrayList<>();
+        visitPairsRecursive(hash, encounteredKeys);
+        return hash;
+    }
 
+    private void visitPairsRecursive(HashParseNode hash, List<ParseNode> encounteredKeys) {
         for (ParseNodeTuple pair : hash.getPairs()) {
             ParseNode key = pair.getKey();
-            if (key == null) continue;
-            int index = encounteredKeys.indexOf(key);
+            if (key == null) {
+                if (pair.getValue() instanceof HashParseNode) {
+                    visitPairsRecursive((HashParseNode) pair.getValue(), encounteredKeys);
+                }
+                continue;
+            }
+            int index = matchesExistingIndex(key, encounteredKeys);
             if (index >= 0) {
-                warn(hash.getPosition(), "key " + key +
+                encounteredKeys.set(index, key);
+                warn(hash.getPosition(), "key " + keyToString(key) +
                         " is duplicated and overwritten on line " + (encounteredKeys.get(index).getPosition().toSourceSection(lexer.getSource()).getStartLine()));
             } else {
                 encounteredKeys.add(key);
             }
         }
+    }
 
-        return hash;
+    private String keyToString(ParseNode node) {
+        if (node instanceof SymbolParseNode) {
+            return ":" + ((SymbolParseNode) node).getName();
+        } else {
+            return node.toString();
+        }
+    }
+
+    private int matchesExistingIndex(ParseNode currentNode, List<ParseNode> encounteredKeys) {
+        int found = -1;
+        for (int i = 0; i < encounteredKeys.size(); i++) {
+            final ParseNode parseNode = encounteredKeys.get(i);
+            if (parseNode instanceof SymbolParseNode && currentNode instanceof SymbolParseNode) {
+                if (((SymbolParseNode) parseNode).getRope().equals(((SymbolParseNode) currentNode).getRope())) {
```

Comment:
Maybe we want to implements `equals` in SymbolParseNode to do this comparison?
That would simplify the logic here.

Reply:
@eregon I did that initially but the objects are seemingly different since they have different source sections.  The literals they contain however are equal.  I think there should be a better abstraction to represent equality here of the literals in the parse nodes but I'm not sure yet exactly how.

Reply:
I think maybe a different equality method like `valueEquals()` defined on an interface for literal nodes could be a nice way to model this.

Reply:
Could I add a TODO for this to resolve in a later time?

Reply:
OK, but since it seems no node implements `equals()` and they cannot be `==`, we might as well only handle SymbolParseNode.

Reply:
@eregon I've update to only handle `SymbolParseNode` and added the `TODO`